atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTrackerTest.java

```diff
@@ -0,0 +1,306 @@
+                        TABLE_2,
+                                createTransactionReadInfo(
+                                        BYTES_READ_1 + 2 * BYTES_READ_2 + BYTES_READ_3,
+                                        3,
+                                        createMaximumBytesKvsCallInfo(BYTES_READ_3, KVS_METHOD_NAME_3)),
+                        TABLE_3,
+                                createTransactionReadInfo(
+                                        2 * BYTES_READ_1 + 2 * BYTES_READ_2,
+                                        2,
+                                        createMaximumBytesKvsCallInfo(BYTES_READ_2, KVS_METHOD_NAME_1))));
+    }
+
+    /**
+     * Schedules tracking calls from {@link #differentMethodCallsAreTracked} repeatedly ({@link #CONCURRENCY_ROUNDS}
+     * times.
+     */
+    @Test
+    public void concurrencyAndOrderOfExecutionTest() throws InterruptedException {
+        ExecutorService executor = PTExecutors.newFixedThreadPool(100);
+
+        for (int round = 0; round < CONCURRENCY_ROUNDS; round++) {
+            executor.execute(() -> tracker.recordReadForTable(TABLE_1, KVS_METHOD_NAME_1, BYTES_READ_1));
+            executor.execute(() -> tracker.recordReadForTable(TABLE_1, KVS_METHOD_NAME_2, BYTES_READ_3));
+            executor.execute(() -> tracker.recordReadForTable(TABLE_1, KVS_METHOD_NAME_3, BYTES_READ_3));
+            executor.execute(() -> {
+                Consumer<Long> tableOneFirstTracker = tracker.recordCallForTable(TABLE_1);
+                tableOneFirstTracker.accept(BYTES_READ_3);
+            });
+            executor.execute(() -> {
+                Consumer<Long> tableOneSecondTracker = tracker.recordCallForTable(TABLE_1);
+                tableOneSecondTracker.accept(BYTES_READ_2);
+            });
+
+            executor.execute(() -> tracker.recordReadForTable(TABLE_2, KVS_METHOD_NAME_2, BYTES_READ_2));
+            executor.execute(() -> tracker.recordReadForTable(TABLE_2, KVS_METHOD_NAME_3, BYTES_READ_3));
+            executor.execute(() -> {
+                Consumer<Long> tableTwoTracker = tracker.recordCallForTable(TABLE_2);
+                tableTwoTracker.accept(BYTES_READ_1);
+                tableTwoTracker.accept(BYTES_READ_2);
+            });
+
+            executor.execute(() -> tracker.recordReadForTable(TABLE_3, KVS_METHOD_NAME_1, BYTES_READ_2));
+            executor.execute(() -> {
+                Consumer<Long> tableThreeTracker = tracker.recordCallForTable(TABLE_3);
+                tableThreeTracker.accept(BYTES_READ_1);
+                tableThreeTracker.accept(BYTES_READ_1);
+                tableThreeTracker.accept(BYTES_READ_2);
+            });
+
+            executor.execute(() -> tracker.recordTableAgnosticRead(KVS_METHOD_NAME_4, BYTES_READ_1));
+            executor.execute(() -> tracker.recordTableAgnosticRead(KVS_METHOD_NAME_5, BYTES_READ_2));
+
+            executor.execute(tracker::getReadInfo);
```

Comment:
note - you do this to see if there's concurrent modification exceptions, but the exceptions are swallowed so you'll never actually test that :)

As discussed offline, see LockWatchEventLogTest for inspiration